fs2/src/main/scala/laserdisc/fs2/RedisConnection.scala

```diff
@@ -38,26 +38,44 @@ object RedisConnection {
     }
 
   private[fs2] final object impl {
+
     def send[F[_]: Applicative](sink: Sink[F, Byte])(implicit log: LogWriter[F]): Sink[F, RESP] =
       _.evalMap(resp => log.debug(s"sending $resp") *> resp.pure)
         .through(streamEncoder.encode)
         .flatMap(bits => Stream.segment(Segment.array(bits.toByteArray)))
         .to(sink)
 
     def receive[F[_]: Effect](implicit log: LogWriter[F]): Pipe[F, Byte, RESP] = {
-      def toBitVector: Pipe[F, Byte, BitVector] = {
-        def go(stream: Stream[F, Byte]): Pull[F, BitVector, Unit] = stream.pull.unconsChunk.flatMap {
-          case Some((chunk, rest)) => Pull.output1(BitVector.view(chunk.toByteBuffer)) >> go(rest)
-          case _                   => Pull.done
-        }
-
-        stream =>
-          go(stream).stream
+
+      def toBitVector: Pipe[F, Byte, NonEmptyFrame] = {
+
+        def go(stream: Stream[F, Byte], previous: Frame): Pull[F, NonEmptyFrame, Unit] =
+          stream.pull.unconsChunk flatMap {
+
+            case Some((chunk, rest)) =>
```

Comment:
here, instead of handling the chunk directly I delegate the frame to decide if the vector is complete according to the RESP protocol.